File: app/src/main/java/com/zhoujiulong/simpleprojectframework/base/BasePresenter.java
package com.zhoujiulong.simpleprojectframework.base;


import com.zhoujiulong.baselib.http.HttpUtil;

public abstract class BasePresenter<M extends BaseModel, T extends BaseView> {

    protected M mModel;
    protected T mView;
    /**
     * 網絡請求標記tag
     */
    protected String ReTag = getClass().getName();

    public BasePresenter() {
        initModel();
    }

    public abstract void initModel();

    public void attachView(T View, String ReTag) {
        this.mView = View;
        this.ReTag = ReTag;
        if (mModel != null) {
            mModel.attachView(ReTag);
        }
    }

    public void detachView() {
        HttpUtil.cancelWithTag(ReTag);
        if (mView != null) {
            mView = null;
        }
        if (mModel != null) {
            mModel.detachView();
            mModel = null;
        }
    }

}

File: app/src/main/java/com/zhoujiulong/simpleprojectframework/base/BasePresenter.kt
package com.zhoujiulong.simpleprojectframework.base


import com.zhoujiulong.baselib.http.HttpUtil

abstract class BasePresenter<M : BaseModel<*>, T : BaseView> {

    protected var mModel: M? = null
    protected var mView: T? = null
    /**
     * 網絡請求標記tag
     */
    protected var ReTag = javaClass.name

    init {
        mModel = this.initModel()
    }

    abstract fun initModel(): M

    fun attachView(view: Any, ReTag: String) {
        this.ReTag = ReTag
        @Suppress("UNCHECKED_CAST")
        this.mView = view as? T
        mModel?.attachView(ReTag)
    }

    fun detachView() {
        HttpUtil.cancelWithTag(ReTag)
        mView = null
        mModel?.detachView()
        mModel = null
    }

}
